### PR TITLE
Course Units Schema

### PR DESCRIPTION
**Pull request by mxc-maggiechen**
## Notion ticket link

[Course Units Schema](https://www.notion.so/uwblueprintexecs/Course-Units-Schemas-3127e0e24f1b4208a410a081292691a9?pvs=4)

## Implementation description

- Added new file `backend/models/courseunit.mgmodel.ts` that exports the Course Unit model and its schema
- Course Unit schema looks like the following:

```
id: string
displayIndex: number # e.g. 1 for Unit 1
title: string
modules: [Module]
```

```
id: string
title: string
pages: [Page]
```

```
id: string
title: string
type: "Lesson" | "Activity"
layout: [ElementSkeleton]
```

```
id: string
x: number
y: number
w: number
h: number
content: string # reference to element by id
```

## Steps to test
1.

## What should reviewers focus on?
* Please take a close look at the schema to make sure we didn't mess anything up!

## Checklist
- [x] My PR name is descriptive and in imperative tense
- [x] My commit messages are descriptive and in imperative tense. My commits are atomic and trivial commits are squashed or fixup'd into non-trivial commits
- [x] I have run the appropriate linter(s)
- [x] I have requested a review from the PL, as well as other devs who have background knowledge on this PR or who will be building on top of this PR

### CODE REVIEW

File: backend/models/courseunit.mgmodel.ts [hunk 1]
```diff
@@ -0,0 +1,105 @@
+import mongoose, { Schema, Document } from "mongoose";
+
+type ElementSkeleton = {
```
Comment (carolynzhang18) [inline backend/models/courseunit.mgmodel.ts:3]: Could we `export` the type definitions for `ElementSkeleton`, `Page`, and `Module`? I'm thinking our backend operations will likely also make use of these types once we start implementing our APIs

File: backend/models/courseunit.mgmodel.ts [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+  content: {
+    type: String,
+    required: true,
+  },
+});
+
+type Page = {
+  id: string;
+  title: string;
+  type: "Lesson" | "Activity";
```
Comment (carolynzhang18) [inline backend/models/courseunit.mgmodel.ts:38]: Could we move `"Lesson" | "Activity"` out into its own type?
e.g. `export type PageType = "Lesson" | "Activity";`

We could have this type in `backend/types/courseTypes.ts` (you'll need to do a `git pull origin main` to see this `backend/types/` folder, I just added it yesterday 😅)

File: backend/models/courseunit.mgmodel.ts [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+  layout: [ElementSkeleton];
+};
+
+const PageSchema: Schema = new Schema({
+  title: {
+    type: String,
+    required: true,
+  },
+  type: {
```
Comment (carolynzhang18) [inline backend/models/courseunit.mgmodel.ts:47]: To match the `Page.type` type definition above, we'll want the type validator `String` and an enum validator `["Lesson", "Activity"]` (similar to what we do for `User.role`)

File: backend/models/courseunit.mgmodel.ts [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+  pages: [Page];
+};
+
+const ModuleSchema: Schema = new Schema({
+  id: {
+    type: String,
+    required: true,
+  },
+  title: {
+    type: Number,
```
Comment (carolynzhang18) [inline backend/models/courseunit.mgmodel.ts:69]: type 🧶 :)

**Review by carolynzhang18 — COMMENTED** [review]
A few minor fixes and some type exporting, but overall it's looking good! 🤩

Have you guys tested creating documents with this model yet BTW? If not, you can try creating a hard-coded `CourseUnit` document using Mongoose (https://masteringjs.io/tutorials/mongoose/create) and see if it saves in the DB without problems

**Review by carolynzhang18 — COMMENTED** [review]
Just the missing an s on "modules" in the schema lol everything else looks good to me!! :D

**Review by carolynzhang18 — APPROVED** [review]
image

File: backend/models/courseunit.mgmodel.ts [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+  w: {
+    type: Number,
+    required: true,
+  },
+  h: {
+    type: Number,
+    required: true,
+  },
+  content: {
+    type: String,
```
Comment (achandrabalan) [inline backend/models/courseunit.mgmodel.ts:31]: Content type could be an enum as there should be a fixed number of types of components. Ie "Text Box" | " Match Box" | "Fill in the Blank"

However, maybe as the number of types of components increase in the future this becomes more tedious / messy.

**Review by achandrabalan — APPROVED** [review]
Nailed it
image

File: backend/models/courseunit.mgmodel.ts (hunk 5 above)
Comment (carolynzhang18) [reply backend/models/courseunit.mgmodel.ts:31]: The idea with the `content` property of the `ElementSkeleton` is that it just contains an ID which maps us to a separate collection of `CourseElements` - which is why we're making it a `String`.

The `courseElements` schema is still a WIP haha, but 100% agree we should have an enum for the different types of components when we get there :)

Comment (carolynzhang18) [reply backend/models/courseunit.mgmodel.ts:31]: also LMAO love the papas pizzeria pic

**Pull request merged**